### ISSUE TITLE
decay_rate behaves differently based on batch_size

### ISSUE THREAD

**Issue by carlogrisetti** [ludwig-ai/ludwig#1044]
Seems to me, purely based on an empyrical observation (I did not check within the code), that `decay` (and more specifically `decay_rate`) is computer at a per-step (or per-batch) time, and not over the epoch.

I don't know if this is the intended behavior (decreasing the learning rate within the same epoch, having the first steps\batches using a higher learning rate than the last ones), but it could be, since it's the same logic that governs the `learning_rate_warmup_epochs` but even if this is by-design there is a tricky thing that happens:

If I have a smaller `batch_size` (thus a higher number of batches\steps in an epoch, hence more learning rate decay applied) I will get a higher rate of decay as opposed to the same training using a larger `batch_size` (thus a lower number of batches\steps in an epoch, hence less learning rate decay applied).

Does anything above make sense to you?
If so, this could be fixed in 2 ways:
1) compute the `decay_rate` effect at the epoch level, making it batch size independent
2) compute the `decay_rate` effect at the batch\step level (as it is now), but making its effect proportional to the number of steps in an epoch (the same thing that happens in the `learning_rate_warmup_epochs`)

One more thing: I think that the `decay` behavior should come into effect after the `learning_rate_warmup` has completed stepping up, otherwise we get a "curved" ramp up to the initial `learning_rate` value (that never gets achieved, since it's already decaying the value from the start).

**Comment by carlogrisetti** [comment]
same exact model config with `learning_rate: 0.001` and `decay_rate: 0.90`, but with...

`batch_size: 50`
![image](https://user-images.githubusercontent.com/4464640/101278393-41f63a80-37bb-11eb-8578-83055034e4d2.png)

`batch_size: 50`
![image](https://user-images.githubusercontent.com/4464640/101278388-386cd280-37bb-11eb-9d50-abbf1b9f8962.png)
